src/sentry/releasehealth/metrics.py

```diff
@@ -150,3 +158,185 @@ def _compute_crash_free_rate(data: Dict[str, float]) -> Optional[float]:
+                )
+
+            if not total:
+                release_tag_values = []
+
+                for _, release in project_releases:
+                    value = indexer.resolve(org_id, UseCase.TAG_VALUE, release)  # type: ignore
+                    if value is not None:
+                        # We should not append the value if it hasn't been
+                        # observed before.
+                        release_tag_values.append(value)
+
+                where_common.append(
+                    Condition(Column(tag_key(org_id, "release")), Op.IN, release_tag_values)
+                )
+
+            return where_common
+
+        def _get_common_groupby(total: bool) -> List[SelectableExpression]:
+            if total:
+                return [Column("project_id")]
+            else:
+                return [Column("project_id"), Column(tag_key(org_id, "release"))]
+
+        def _convert_results(data: Any, total: bool) -> Dict[Any, int]:
+            if total:
+                return {x["project_id"]: x["value"] for x in data}
+            else:
+                release_tag = tag_key(org_id, "release")
+                return {(x["project_id"], x[release_tag]): x["value"] for x in data}
+
+        def _count_sessions(total: bool, referrer: str) -> Dict[Any, int]:
+            query = Query(
+                dataset=Dataset.Metrics.value,
+                match=Entity("metrics_counters"),
+                select=[Column("value")],
+                where=_get_common_where(total)
+                + [
+                    Condition(Column("metric_id"), Op.EQ, metric_id(org_id, "session")),
+                ],
+                groupby=_get_common_groupby(total),
+            )
+
+            return _convert_results(
+                raw_snql_query(
+                    query,
+                    referrer=referrer,
+                    use_cache=False,
+                )["data"],
+                total=total,
+            )
+
+        def _count_users(total: bool, referrer: str) -> Dict[Any, int]:
+            query = Query(
+                dataset=Dataset.Metrics.value,
+                match=Entity("metrics_sets"),
+                select=[Column("value")],
+                where=_get_common_where(total)
+                + [
+                    Condition(Column("metric_id"), Op.EQ, metric_id(org_id, "user")),
+                ],
+                groupby=_get_common_groupby(total),
+            )
+
+            return _convert_results(
+                raw_snql_query(
+                    query,
+                    referrer=referrer,
+                    use_cache=False,
+                )["data"],
+                total=total,
+            )
+
+        # XXX(markus): Four queries are quite horrible for this... the old code
+        # sufficed with two. From what I understand, ClickHouse would have to
+        # gain a function uniqCombined64MergeIf, i.e. a conditional variant of
+        # what we already use.
```

Comment:
Sets are tricky, but for the counter we could make only the `sessions_per_release` query, and derive the summed `sessions_per_project` in python code, right?

Reply:
`sessions_per_release` does not necessarily sum up to `sessions_per_project`, as we don't fetch all releases of a project, only the ones the user asked us to.

Reply:
Right, sorry, `AND GIVEN RELEASES` could have given me a hint :)
Then all we could do is remove the filter from the query and do the filtering in python code as well, but let's not get into that unless it's really necessary.

Reply:
I actually have an idea, we could write a query that returns sessions per release, then another row for all other sessions. This:

```
select release, count()
from metrics
where metric_id = 'session'
and tags[session.status] = 'init'
and release in ('a', 'b', 'c')
group by release
```

becomes:

```
select if(release in ('a', 'b', 'c'), release, null) as maybe_release, count()
from metrics
where metric_id = 'session'
and tags[session.status] = 'init'
group by maybe_release
```

then all counts can be summed up.

However, you only save one query for that, as you said sets are tricky.

[There are functions in Snuba that give you back the raw merge state](https://clickhouse.com/docs/en/sql-reference/aggregate-functions/combinators/#aggregate_functions_combinators-mergestate), however it would probably take another query to convert that merge state into something useful anyway.